server.go

```
@@ -1607,104 +1566,83 @@ func (s *Server) WriteSeries(database, retentionPolicy string, points []Point) (
+	for _, p := range points {
+		// Local function makes lock management foolproof.
+		measurement, series, err := func() (*Measurement, *Series, error) {
+			s.mu.RLock()
+			defer s.mu.RUnlock()
+			db := s.databases[database]
+			if db == nil {
+				return nil, nil, fmt.Errorf("database not found %q", database)
+			}
+			if measurement, series := db.MeasurementAndSeries(p.Name, p.Tags); series != nil {
+				return measurement, series, nil
+			}
+			panic("series not found")
+		}()
+		if err != nil {
+			return 0, err
 		}
-	}
-	return index, err
-}
-
-func (s *Server) writePoint(database, retentionPolicy string, point *Point) (uint64, error) {
-	measurement, tags, timestamp, values := point.Name, point.Tags, point.Timestamp, point.Values
-
-	// Sanity-check the data point.
-	if measurement == "" {
-		return 0, ErrMeasurementNameRequired
-	}
-	if len(values) == 0 {
-		return 0, ErrValuesRequired
-	}
 
-	// Find the id for the series and tagset
-	seriesID, err := s.createSeriesIfNotExists(database, measurement, tags)
-	if err != nil {
-		return 0, err
-	}
-
-	// Retrieve measurement.
-	m, err := s.measurement(database, measurement)
-	if err != nil {
-		return 0, err
-	} else if m == nil {
-		return 0, ErrMeasurementNotFound
-	}
+		// Retrieve shard group.
+		g, err := s.shardGroupByTimestamp(database, retentionPolicy, p.Timestamp)
+		if err != nil {
+			return 0, err
+		}
 
-	// Retrieve shard group.
-	g, err := s.createShardGroupIfNotExists(database, retentionPolicy, timestamp)
-	if err != nil {
-		return 0, fmt.Errorf("create shard(%s/%s): %s", retentionPolicy, timestamp.Format(time.RFC3339Nano), err)
-	}
+		// Find appropriate shard within the shard group.
+		sh := g.ShardBySeriesID(series.ID)
 
-	// Find appropriate shard within the shard group.
-	sh := g.ShardBySeriesID(seriesID)
+		// Get a field codec.
+		s.mu.RLock()
+		codec := NewFieldCodec(measurement)
+		s.mu.RUnlock()
+		if codec == nil {
+			panic("field codec is nil")
```

We should really only be panicing in the raft apply loop. we should return an error in this instance.